docs/proposals/distributed-query-execution.md

```diff
@@ -0,0 +1,197 @@
+
+To support this, the querier now implements two interfaces:
+* **QueryExecutor** – Executes assigned query plan root fragment.
+* **QueryServer** – Serves results to other queriers via streaming gRPC endpoints. (i.e., acts as a remote data source for `Remote` operators).
+
+#### QueryExecutor
+The QueryExecutor interface enables a querier to execute the root fragment of the distributed query plan.
+
+```
+type QueryExecutor interface {
+	ExecuteQuery(p LogicalPlan) (QueryResult, error)
+}
+```
+
+#### QueryServer
+The QueryServer interface enables a querier to expose query results to other queriers via a gRPC API. This is used when another querier is executing a fragment that depends on the results of this one (i.e., the child in the execution tree).
+
+```
+type QueryServer interface {
+	Enqueue(context.Context, p LogicalPlan) error
+    Series(SeriesRequest, querierpb.QueryServer_SeriesServer) error
+	Next(*executionpb.NextRequest, querierpb.QueryServer_NextServer) error
+}
+```
+
+* **Enqueue** - Schedules the provided logical plan fragment for execution on the querier. Enqueuing a plan on a quereir will block a concurrency until the query finishes, or times out if it exceeds the configured limit. It respects the querier’s concurrency limits and queue capacity.
+* **Series** - Streams the metadata (labels) for the series produced by the fragment, allowing parent queriers to understand the shape of incoming data.
+* **Next** - Streams the actual step vector results of the fragment, returning time-step-wise computed samples and histograms.
+
+These methods are implemented as streaming gRPC endpoints and used by `Remote` operators to pull data on-demand during execution.
+
+
+##### gRPC Service Definition
+
+```
+service QueryServer {
+  rpc Series(SeriesRequest) returns (stream OneSeries);
+  rpc Next(NextRequest) returns (stream StepVectorBatch);
+}
+
+message OneSeries {
+    repeated Label labels = 1;
+}
+
+message StepVectorBatch {
+    repeated StepVector step_vectors = 1;
+}
+
+message StepVector {
+	int64 t  = 1;
+	repeated uint64 sample_IDs = 2;
+	repeated double samples = 3;
+
+	repeated uint64 histogram_IDs = 4;
+	repeated FloatHistogram histograms = 5;
+}
+```
+
+Together, the QueryExecutor and QueryServer roles allow each querier to act as both a compute node and a data provider, enabling true parallel and distributed evaluation of PromQL queries across Cortex's infrastructure.
+
+#### Fragment Execution
+
+Each fragment is evaluated using the **Thanos PromQL engine’s volcano iterator model**, where operators are chained and lazily pulled for evaluation. If a fragment includes a `RemoteOperator`, it fetches the intermediate result from the peer querier serving the dependent fragment. Leaf operators such as `MatrixSelector` and `VectorSelector` read data from storage layer.
+
+#### Co-ordinated Execution
+
+The querier assigned the **root fragment** also plays the role of **coordinator**. The co-ordinator will
```

Comment:
The root querier will call `Series` and `Next` to the child querier. Those requests will also specify the logical plan ID or some sort of request ID so the child querier finds out the actual plan?

Reply:
Good point. Yes, we need to refer fragments by ID.

Reply:
I think we don't need to include that detail in the proposal and leave it up to implementation.